### PR TITLE
Action Creator 8: Required Fields and Default Values

### PR DESCRIPTION
**Pull request by iethree**
## Description

This allows form fields in action forms to be set as required. If they are not required, this allows a user to set a default value for that field.

## Demo

https://www.loom.com/share/ed58b57f729c48a69f03ee74c2326d87

## Testing Steps

- create a new action with at least 2 different form fields
- fields should be required by default, but mark one field as non-required and provide a default value for it.
- add that action to an app page (add a button, add a click action to your newly-created button and make it link to that particular action)
- exit edit mode and click your button to get an action form
- see that the save button is disabled until all required fields have values
- leave your non-required field blank and see that it saves your default value
- fill in your non-required field and see that it sets the provided (rather than the default) value

### CODE REVIEW

File: frontend/src/metabase/writeback/utils.ts [hunk 1]
```diff
@@ -67,7 +67,7 @@ export const isActionButtonCard = (card: SavedCard) =>
 export function isActionButtonDashCard(
   dashCard: BaseDashboardOrderedCard,
 ): dashCard is ActionButtonDashboardCard {
-  const virtualCard = dashCard.visualization_settings?.virtual_card;
+  const virtualCard = dashCard?.visualization_settings?.virtual_card;
   return isActionButtonCard(virtualCard as SavedCard);
 }
```
Comment (kulyk) [inline frontend/src/metabase/writeback/utils.ts:72]: Curios what should happen so a null-ish `dashCard` gets passed here? Might make sense to also type it like `dashCard?: BaseDashboardOrderedCard` to enforce type-safety

File: frontend/src/metabase/lib/validate.js [hunk 2]
```diff
@@ -19,6 +23,7 @@ function makeValidate(steps = []) {
   function all(...args) {
     return steps.map(step => step(...args)).filter(e => e);
   }
+  validate.required = undefined; // to help typescript out
```
Comment (kulyk) [inline frontend/src/metabase/lib/validate.js:26]: TypeScript appreciates the help 😊

File: frontend/src/metabase/writeback/containers/ActionParametersInputForm/utils.unit.spec.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
+import type { FieldSettingsMap } from "metabase-types/api";
+import { formatParametersBeforeSubmit, setDefaultValues } from "./utils";
+
+describe("writeback > containers > ActionParametersInputForm > utils", () => {
+  describe("formatParametersBeforeSubmit", () => {
```
Comment (kulyk) [inline frontend/src/metabase/writeback/containers/ActionParametersInputForm/utils.unit.spec.ts:5]: Thank you sooo much for covering all that! ❤️

**Review by kulyk — APPROVED** [review]
Nice work! Noticed a few things, but they're absolutely non-blocking

1. Seems like the default value is saved as a string all the time even if the field type is set to `number`
2. If I create an action with a couple of parameters, save it without touching parameter settings at all, map it to a button and click it, I'd get a FE crash about `placeholder` being `undefined`

Captured both in the video down below:

https://user-images.githubusercontent.com/17258145/190638174-bfcd26d6-34ee-414d-9ccf-f16cac6d6e9c.mp4

File: frontend/src/metabase/writeback/containers/ActionParametersInputForm/utils.unit.spec.ts (hunk 3 above)
Comment (iethree) [reply frontend/src/metabase/writeback/containers/ActionParametersInputForm/utils.unit.spec.ts:5]: copilot wrote the this test actually 😂  it's really good at writing simple tests.

File: frontend/src/metabase/writeback/utils.ts (hunk 1 above)
Comment (iethree) [reply frontend/src/metabase/writeback/utils.ts:72]: not sure how this happened, but I was running into crashes here, maybe on an initial render it was sending through a null dashcard?

**Pull request merged**